crates/uv-requirements/src/discovery.rs

```diff
@@ -0,0 +1,578 @@
+                    err
+                );
+                return;
+            }
+        };
+
+        if let Some(workspace) = pyproject_toml
+            .tool
+            .as_ref()
+            .and_then(|tool| tool.uv.as_ref())
+            .and_then(|uv| uv.workspace.as_ref())
+        {
+            let is_excluded = match is_excluded_from_workspace(
+                workspace,
+                outer_workspace_root,
+                inner_workspace_root,
+            ) {
+                Ok(contents) => contents,
+                Err(err) => {
+                    warn_user!(
+                        "Invalid pyproject.toml `{}`: {}",
+                        pyproject_toml_path.user_display(),
+                        err
+                    );
+                    return;
+                }
+            };
+            if !is_excluded {
+                warn_user!(
+                    "Outer workspace including existing workspace, nested workspaces are not supported: `{}`",
+                    pyproject_toml_path.user_display(),
+                );
+            }
+        }
+
+        // We're in the examples or tests of another project (not a workspace), this is fine.
+        return;
+    }
+}
+
+fn is_excluded_from_workspace(
+    workspace: &ToolUvWorkspace,
+    workspace_root: &Path,
+    project_path: &Path,
+) -> Result<bool, DiscoverError> {
+    // Check if we're in the excludes of a workspace.
+    for exclude_glob in workspace.exclude.iter().flatten() {
+        let absolute_glob = workspace_root
+            .join(exclude_glob.as_str())
+            .to_string_lossy()
+            .to_string();
+        for excluded_root in glob(&absolute_glob)
+            .map_err(|err| DiscoverError::Pattern(absolute_glob.to_string(), err))?
+        {
+            let excluded_root =
+                excluded_root.map_err(|err| DiscoverError::Glob(absolute_glob.to_string(), err))?;
+            if excluded_root == project_path {
+                return Ok(true);
+            }
+        }
+    }
+    Ok(false)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::env;
+    use std::path::Path;
+
+    use insta::assert_json_snapshot;
+
+    use crate::discovery::ProjectWorkspace;
+
+    fn workspace_test(folder: impl AsRef<Path>) -> (ProjectWorkspace, String) {
+        let root_dir = env::current_dir()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("scripts")
+            .join("workspaces");
+        let project = ProjectWorkspace::discover(root_dir.join(folder)).unwrap();
+        let root_escaped = regex::escape(root_dir.to_string_lossy().as_ref());
+        (project, root_escaped)
+    }
+
+    #[test]
+    fn albatross_in_example() {
+        let (project, root_escaped) = workspace_test("albatross-root-workspace");
+        let filters = vec![(root_escaped.as_str(), "[ROOT]")];
+        insta::with_settings!({filters => filters}, {
```

Comment:
I tried the json snapshots here for the first time and find them better than the regular debug snapshots